File: mifos-passcode/src/main/java/com/mifos/mobile/passcode/BasePassCodeActivity.java
package com.mifos.mobile.passcode;

import com.mifos.mobile.passcode.utils.ForegroundChecker;

import androidx.activity.ComponentActivity;


/**
 * Created by dilpreet on 19/01/18.
 */

public abstract class BasePassCodeActivity extends ComponentActivity implements
        ForegroundChecker.Listener {

    @Override
    protected void onResume() {
        super.onResume();
        ForegroundChecker.get().addListener(this);
        ForegroundChecker.get().onActivityResumed();
    }

    @Override
    protected void onPause() {
        super.onPause();
        ForegroundChecker.get().onActivityPaused();
    }

    @Override
    public void onBecameForeground() {
        MifosPassCodeActivity.startMifosPassCodeActivity(this, getPassCodeClass(),
                false);
    }

    public abstract Class getPassCodeClass();

}

File: mifos-passcode/src/main/java/com/mifos/mobile/passcode/BasePassCodeActivity.kt
package com.mifos.mobile.passcode

import androidx.activity.ComponentActivity
import androidx.appcompat.app.AppCompatActivity
import com.mifos.mobile.passcode.utils.ForegroundChecker
import com.mifos.mobile.passcode.utils.ForegroundChecker.Companion.get

/**
 * Created by dilpreet on 19/01/18.
 */
abstract class BasePassCodeActivity : ComponentActivity(), ForegroundChecker.Listener {
    override fun onResume() {
        super.onResume()
        get()!!.addListener(this)
        get()!!.onActivityResumed()
    }

    override fun onPause() {
        super.onPause()
        get()!!.onActivityPaused()
    }

    override fun onBecameForeground() {
        MifosPassCodeActivity.startMifosPassCodeActivity(
            this, passCodeClass,
            false
        )
    }

    abstract val passCodeClass: Class<*>?
}